reproducible_build.sh

```diff
@@ -56,74 +80,258 @@ if [ "$CLONE" = false ]; then
+    set -u
+
+    build_dependency "${BOOST_PREFIX}" "${LOG_DIR}/boost_build.log" build_boost
+    build_dependency "${LIBEVENT_PREFIX}" "${LOG_DIR}/libevent_build.log" build_libevent
+}
+
+function build_bdb() {
+    wget http://download.oracle.com/berkeley-db/db-4.8.30.NC.tar.gz
+    tar xf db-4.8.30.NC.tar.gz
+    if [ ${OS_NAME} = "osx" ]; then
+	curl -OL https://raw.github.com/narkoleptik/os-x-berkeleydb-patch/master/atomic.patch
+	patch db-4.8.30.NC/dbinc/atomic.h < atomic.patch
+    fi
+    cd db-4.8.30.NC/build_unix
+    BDB_LOG="${LOG_DIR}/bdb_build.log"
+    echo "Building bdb.  tail -f $BDB_LOG to see the details and monitor progress"
+    ../dist/configure --prefix=$BDB_PREFIX --enable-cxx --disable-shared --with-pic > "${BDB_LOG}"
+    background make "${BDB_LOG}" "Waiting for bdb to finish building"
+    make install >> "${BDB_LOG}" 2>&1
+}
+
+function build_openssl() {
+    OPENSSL_LOG="$1"
+    wget https://www.openssl.org/source/openssl-1.0.1p.tar.gz
+    tar xf openssl-1.0.1p.tar.gz
+    mkdir -p $OPENSSL_PREFIX/ssl
+    cd openssl-1.0.1p
+    echo "Building bdb.  tail -f $OPENSSL_LOG to see the details and monitor progress"
+    if [ ${OS_NAME} = "osx" ]; then
+	./Configure --prefix=$OPENSSL_PREFIX --openssldir=$OPENSSL_PREFIX/ssl \
+		    -fPIC darwin64-x86_64-cc \
+		    no-shared no-dso no-engines > "${OPENSSL_LOG}"
+    else
+	./Configure --prefix=$OPENSSL_PREFIX --openssldir=$OPENSSL_PREFIX/ssl \
+		    linux-x86_64 -fPIC -static no-shared no-dso > "${OPENSSL_LOG}"
+    fi
+    background make "${OPENSSL_LOG}" "Waiting for openssl to finish building"
+    make install >> "${OPENSSL_LOG}" 2>&1
+}
+
+function build_boost() {
+    BOOST_LOG="$1"
+    wget http://sourceforge.net/projects/boost/files/boost/1.59.0/boost_1_59_0.tar.bz2/download \
+	 -O boost_1_59_0.tar.bz2
+    tar xf boost_1_59_0.tar.bz2
+    cd boost_1_59_0
+    echo "Building Boost.  tail -f ${BOOST_LOG} to see the details and monitor progress"
+    ./bootstrap.sh --prefix=${BOOST_PREFIX} > "${BOOST_LOG}" 2>&1
+    background "./b2 link=static cxxflags=-fPIC install" \
+	       "${BOOST_LOG}" \
+	       "Waiting for boost to finish building"
+}
+
+function build_libevent() {
+    LIBEVENT_LOG="$1"
+    git clone https://github.com/libevent/libevent.git
+    cd libevent
+    echo "Building libevent.  tail -f ${LIBEVENT_LOG} to see the details and monitor progress"
+    ./autogen.sh > "${LIBEVENT_LOG}" 2>&1
+    ./configure --prefix="${LIBEVENT_PREFIX}" --enable-static --disable-shared --with-pic \
+		LDFLAGS="-L${OPENSSL_PREFIX}/lib/" \
+		CPPFLAGS="-I${OPENSSL_PREFIX}/include" >> "${LIBEVENT_LOG}" 2>&1
+    background make "${LIBEVENT_LOG}" "Waiting for libevent to finish building"
+    make install >> "${LIBEVENT_LOG}"
+}
+
+function build_dependency() {
+    PREFIX=$1
+    LOG=$2
+    BUILD=$3
+    if [ ! -d "${PREFIX}" ]; then
+	trap "cleanup \"${PREFIX}\" \"${LOG}\"" INT TERM EXIT
```

Comment:
if something fails during the build, we need to delete the build directory or else it will get added to the cache and then we'll never successfully get the dependency built